### ISSUE TITLE
Reporting of incomplete quantifiers in corral

### ISSUE THREAD

**Issue by ryzhyk** [boogie-org/corral#17]
Hello corral developers,  

First off, thank you for the amazing tool!  We've been using corral in SDN verification for a few months now and it's been working really well for us.  

One issue we ran into recently is related to the handling of quantifiers. For example, on the following simple example, corral returns a spurious counterexample trace:

```
function h(x: int): bool;
function f(y: int): int;
function g(x: int): int;

axiom (forall y: int, x: int :: (x == f(y)) ==> (h(x) && (y == g(x))));

//axiom (forall y: int, x: int :: ((x == f(y)) ==> (g(x) == y)));
//axiom (forall y: int, x: int :: ((x == f(y)) ==> h(x)));

procedure main ()
{   
    var y: int;
   
    assert (y == g(f(y)));
}
```

However, if I replace the assertion with two equivalent commented out assertions then corral correctly reports that "Program has no bugs". I am guessing that in the former case z3  fails to validate the spec due to incomplete quantifiers.  I understand that quantifiers are fundamentally hard and that the handling of quantifiers in corral/z3 will always be "best effort".  However, I think that there are a few things corral and z3 could do to handle such situations better:

1. First, it would save me many hours of work if corral reported such situations as "could not verify" rather than returning a spurious trace. I could then try to re-formulate the spec in a more z3-friendly way, as in the above example.
2. It would be even better if corral/z3 could point to the specific axiom causing the problem.
3. In this particular case, the re-writing needed to successfully verify the spec is trivial and perhaps could be done automatically in z3.

I am using the latest corral (commit 89bea8ad106cd052e4c6d774a25f874aef377d37) and z3 (commit f2a869fb5807f01e007c8ac55a7076b72e6628db) from github on Ubuntu 16.04.

Thank you!
Leonid

/cc @NikolajBjorner

**Comment by akashlal** [comment]
@ryzhyk 
Hi Leonid, 
I am happy to know that you're using corral. I would like to help you out here, but I'm not entirely sure how. If you have suggestions (even at an abstract level) I'll be willing to try them out. 

At the Z3 level, the contract is clear. When Z3 isn't sure if its found a model, it returns "unknown". Thus, Corral knows that when Z3 returns SAT, it has indeed found a counterexample in the input program. The trouble is that almost all of the time Z3 returns unknown to Corral. If I go by this, Corral will always report "unsure". This happens because Boogie's VC generation creates many quantifiers, even if the Boogie program doesn't appear to have any. Some of these are because of encoding arrays using Select/Update axioms, some others reflect Boogie's background theory. I haven't exactly traced down all sources of quantifiers.

Perhaps the first step is to nail down all sources of quantifiers. The second step would be to have an algorithm, perhaps at the Z3 level, which can tell what quantifiers resulted in the "unknown". 

Any ideas?

Akash

**Comment by akashlal** [comment]
You can see the VC generated by passing the flag /proverLog:log to corral. Then look at all the log* files.

**Comment by ryzhyk** [comment]
Hi Akash,

Many thanks for the explanation!  I did not realize that unknowns were so common in corral. Can this also lead to corral incorrectly reporting the program as correct? 

Unfortunately I do not have a good solution (or the expertise to suggest one). @NikolajBjorner, who is collaborating with us on this project, earlier pointed out that corral runs z3 with model-based quantifier instantiation disabled. I noticed that enabling MBQI in SMT log files produced by corral is often sufficient for z3 to solve previously "unknown" instances, sometimes without much slowdown.  Perhaps, one strategy would be to try re-running z3 with MBQI enabled (possibly with a timeout) before returning a potentially spurious trace. It would be great to have this available as a command line option.  Alternatively, a more generic interface for controlling z3 settings would do the trick.

Thanks again for looking into this issue!
Leonid

**Comment by akashlal** [comment]
You can use the flag /z3opt:smt.mbqi=true to always turn ON MBQI in Z3. 
Meantime, I'll try to think of a better solution.

**Comment by ryzhyk** [comment]
Sorry about the delay, I wan only now able to try this out.  Unfortunately, the /z3opt:smt.mbqi=true option did not work for me. The generated log files still have MBQI set to false ans z3 returns unknowns.

**Comment by akashlal** [comment]
Sorry, the flag is: /z3opt:SMT.MBQI=true   
The case matters.
For your example, adding this flag makes corral take too much time. When I additionally use the flag /staticInlining then it terminates quickly and reports "no bugs".

**Comment by ryzhyk** [comment]
Thanks, Akash! I am afraid there is still something weird going on with the MBQI option.  When I apply it to a large example where corral previously successfully found a bug, it now reports no bugs on the very first iteration, i.e., while only tracking the assertsPassed variable.  In fact, it seems to do this for all nontrivial specs. When I run z3 directly on the proverLog file, it returns "unsat" with MBQI, and "unknown" without it. I will work on producing a minimal failing example, but first wanted to check with you if there is any known reason MBQI could lead to false negatives and whether false negatives are generally possible with corral?

Thanks!
Leonid

@akashlal

**Comment by ryzhyk** [comment]
Please ignore my previous comment.  Turns out, my automatically generated Boogie file contained an axiom with a quantifier, that could be instantiated to false.

**Comment by akashlal** [comment]
@ryzhyk 
I added a new flag. Supply /explainQuantifiers:f.smt2 to corral. This does two things:
1. Corral will complain if the input file has things that inject unexpected quantifiers into the Z3 query (e.g., if you are using maps in the program but not using Z3's array theory). Corral's error message will guide you how to revise the program.  
2. If Corral reports a bug, it will also dump the file "f.smt2". Give this to Z3. If Z3 says "unknown" (as opposed to "sat") then you cannot trust the result. You can then debug your quantified axioms at the Z3 level using this file, e.g., by asking Nikolaj :) This query should be quick. It encodes a single execution of the input program.

This is a bit hacky, but I hope it saves you some time.

Akash

**Comment by ryzhyk** [comment]
@akashlal , this is really useful, thanks! I have tried the new features.  I am not getting any warning messages (my program does not use maps), but I am indeed getting an "unknown" SMT2 trace, which is really helpful. Unfortunately, I have discovered that both genuine and spurious bugs in my specs manifest in "unknown" traces, but, as you pointed out, this is something we need to investigate.

Thanks,
Leonid
